src/v/cluster/controller_backend.cc

```diff
@@ -1316,29 +1365,77 @@ ss::future<std::error_code> controller_backend::cancel_replica_set_update(
                 errc::waiting_for_recovery);
           }
 
-          const auto raft_cfg_update_finished
+          const auto raft_not_reconfiguring
             = current_cfg.get_state() == raft::configuration_state::simple;
+          // TODO: emit revert command when we have learners in an old
+          // configuration of raft group as they are about to be removed.
 
+          const auto not_yet_moved = are_configuration_replicas_up_to_date(
+            current_cfg, replicas);
+          const auto already_moved = are_configuration_replicas_up_to_date(
+            current_cfg, previous_replicas);
           // raft already finished its part, we need to move replica back
-          if (raft_cfg_update_finished) {
-              return revert_configuration_update(
-                ntp,
-                replicas,
-                replica_revisions,
-                rev,
-                std::move(p),
-                _members_table.local());
+          if (raft_not_reconfiguring) {
+              // move hasn't yet requested
+              if (not_yet_moved) {
+                  // just update configuration revision
+                  auto it = _topics.local().all_topics_metadata().find(
```

Comment:
so the difference between the code in this branch and `update_partition_replica_set` is that the latter is wrapped in `apply_configuration_change_on_leader`, but why do we do it in the force-abort case and not in the soft cancel case?

Reply:
Whole `cancel_replica_set_update` is executed on the leader. Force cancellation should work even when there is no leader. But when configuration change already finished there must be a leader as it was successful, hence we dispatch this last update piece on the leader